### PR TITLE
node: refactor processor

### PR DESCRIPTION
**Pull request by hendrikhofstadt**
This change refactors the core processor to be able to do consensus on entirely different types of payloads implementing Observation. It clearly breaks up the processor into individual state machines.

[](https://mermaid.live/edit#pako:eNqlUztvgzAQ_iuWxyosHRk6Ve1Upc8pZLjiCzkJ2-nZRm2j_PcCLqWhNCCFwcLH9zh_h_cytwplKp0Hj9cEBYNOqsvMiPpZXaxFklyJ5atDrlCl4g6dgwKFbSvgyZoI7SAD_CPmSBWKIgArAiMcFQZ8YByhPQTLQZ8iiRUj5FsyhXhrwesoE5mtyA0ZKOmzMY_VW4Yc75HJKoHvO2JUf0jTziPtPpNGtQy-M2r2Nvhjk-MgX4ydHU2PnTuDAeMMs3igeQL_zWQg2afVf_hO7Pc4zmhkUuZEdKOsyZbj-vPHtaRm1LHe0buyXEiNrIFUfd_2DSaTfosaM5nWrwo3EEqfycwcaigEb58-TC5TzwEXMuxUf0NluoHS4eELNf9LdA)

These reactors handle consensus for a single observable. They are automatically spawned and managed by a `manager`.

For payload/observable specific logic see the `processor.go` that implements VAA specific logic such as storing and broadcasting signed VAAs.

The race detection once again caused issues. I'm starting to think this is a runtime issue.

### CODE REVIEW

**Comment by bruce-riley** [comment]
Why are we removing --devNumGuardians? It is sometimes useful to have that option in tilt.

**Comment by hendrikhofstadt** [comment]
> Why are we removing --devNumGuardians? It is sometimes useful to have that option in tilt.

I just removed the flag in the guardian which was not used anymore. You can still use the tilt `--num` flag.

**Comment by bruce-riley** [comment]
Any concern about the sheer number of reactors given thirty pyth messages per second?

File: node/pkg/processor/reactor/reactor.go [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+)
+
+// ConsensusReactor implements the full consensus processor for a single Observation. It cannot be reused after being
+// finalized.
+type ConsensusReactor[K Observation] struct {
+	// group is the name of the reactor group
+	group string
+	// Current state of the reactor
+	state      State
+	stateMutex sync.Mutex
```
Comment (jynnantonix) [inline node/pkg/processor/reactor/reactor.go:28]: This field is called `stateMutex` but it appears to be protecting many other fields in the struct.  I think it might be a bit clearer if turn `State` into a full struct and put all the mutable fields in there and only keep the immutable fields in the outer `ConsensusReactor`.  Then we can even embed the mutex in the `State` struct so that it's very clear exactly what the mutex is protecting.

**Review by jynnantonix — COMMENTED** [review]
I think the overall design is great and most of my comments are minor issues.

One thing that does make me a bit nervous is the way we're currently using mutexes:

* It's not always clear which functions can be called from multiple goroutines and which functions will be called from a single goroutine.  We should at least annotate the functions that are not thread-safe.  Similarly functions that expect the caller to hold a mutex should say so in their docstrings.
* Manually unlocking mutexes is not panic-safe.  If a goroutine panics while holding a lock and there is no `defer mu.Unlock()` line before the panic then that lock will never be released.  This is why it's generally recommended to put critical sections inside functions and to unlock using `defer`.  In the extreme case we can even use an anonymous function to keep the code inline:

```go
func (s *State) Foo() error {
	// Do some work before taking the lock.

	// define a function for the critical section.
	criticalSection := func() error {
		s.Lock()
		defer s.Unlock()

		// Do work while holding the lock that might panic.

		return nil
	}
	if err := criticalSection(); err != nil {
		return err
	}

	// Do work after releasing the lock.

	return nil
}
```

* The locking discipline needs to be documented since we have locks in a few different places and there are times where we need to acquire multiple locks.  The discipline the code appears to be following is:

  1. `Manager.reactorsLock` before `Manager.gsLock`
  1. `Manager.gsLock` before `Reactor.stateLock`
  1. `Reactor.stateLock` before `Manager.reactorsLock`

	These rules are fine individually but it's clear that there is no deadlock-free way to acquire all three locks at the same time.  We should have these rules documented somewhere so that someone who needs to change this code in the future understands the lock ordering.  Replacing `Manager.gsLock` with an `atomic.Pointer` would also make things a bit easier.

File: node/pkg/processor/vaa_processor.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+			if existing, err := p.getSignedVAA(*db.VaaIDFromVAA(k.CreateVAA(0))); err == nil {
+				p.logger.Debug("ignoring observation since we already have a quorum VAA for it",
+					zap.Stringer("emitter_chain", k.EmitterChain),
+					zap.Stringer("emitter_address", k.EmitterAddress),
+					zap.Uint32("nonce", k.Nonce),
+					zap.Stringer("txhash", k.TxHash),
+					zap.Time("timestamp", k.Timestamp),
+					zap.String("message_id", existing.MessageID()),
+				)
+				continue
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:123]: It does get checked and queued up in the governor in L127.
For existing signed VAAs that should not matter, should it? Otherwise we'd need also pass VAAs received from others to the governor in `handleSignedVAA`

File: node/pkg/processor/reactor/reactor.go [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+	case StateUnobserved:
+		c.lastObservation = time.Now()
+		c.stateTransition(StateObserved)
+	case StateQuorumUnobserved:
+		reactorQuorum.WithLabelValues(c.group, "quorum").Inc()
+		c.stateTransition(StateQuorum)
+	}
+
+	// If we haven't reached quorum in this event, there is nothing more to do
+	if !c.hasQuorum() && c.state != StateQuorum {
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go:413]: This is supposed to prevent the state transition from being triggered twice in the case that L409 transitioned to `StateQuorum` already. 

I'll make this more clear by just adding a return there.

File: node/pkg/processor/reactor/reactor.go [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+	case StateObserved:
+		if time.Since(c.lastObservation) > c.config.QuorumTimeout {
+			c.logger.Debug("timing out", zap.String("reason", "quorum_timeout"))
+			// Time out
+			c.timeOut()
+		}
+
+		if time.Since(c.lastTransmission) > c.config.RetransmitFrequency {
+			// TODO backoff when transmission fails
+			c.logger.Debug("retransmitting", zap.String("reason", "unobserved_timeout"))
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go:240]: removed

File: node/pkg/processor/reactor/reactor.go [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+			reactorResubmission.WithLabelValues(c.group).Inc()
+			// Retransmit signature
+			err := c.transmitSignature(ctx)
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go:243]: Reobservation requests should happen on the layer above I think (i.e. the VAA processor) as they are highly payload specific.
I agree that retransmissions are only slightly useful if the transport layer is unreliable and a previous observation was missed.

File: node/pkg/processor/vaa_processor.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	"google.golang.org/protobuf/proto"
+)
+
+type VAAReactor struct {
+	// lockC is a channel of observed emitted messages
+	lockC chan *common.MessagePublication
+	// obsC is a channel of VAAs used to forward to the reactor manager
+	obsC chan *vaa.VAA
+	// obsvReqSendC is a send-only channel of outbound re-observation requests to broadcast on p2p
+	obsvReqSendC chan<- *gossipv1.ObservationRequest
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:31]: Would keep this here and either add reobservations in this or a follow-up change.

File: node/pkg/processor/reactor/manager.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+				reactorNum.WithLabelValues(p.group).Set(float64(len(p.reactors)))
+			}
+			p.reactorsLock.Unlock()
+			r.ObservationChannel() <- k
+		case m := <-p.confirmationC:
+			digest := ethcommon.BytesToHash(m.Hash)
+
+			p.gsLock.Lock()
+			err := verifySignedObservation(p.group, m, p.gs)
+			p.gsLock.Unlock()
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:134]: I decided to keep the duplicate verification here because we don't want to create reactors for invalid signed observations (potential DoS).
I previously had a way to signal to the reactor that a signed observation was verified before but thought that increased the risk of someone using the code wrong.

File: node/pkg/processor/reactor/manager.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+	// confirmationC is a channel of inbound decoded observations from p2p
+	confirmationC <-chan *gossipv1.SignedObservation
+
+	// setC is a channel of guardian set updates
+	setC <-chan *common.GuardianSet
+
+	// reactors are the current live reactors. This list may contain prepared reactors in StateInitialized
+	reactors     map[ethcommon.Hash]*ConsensusReactor[K]
+	reactorsLock sync.Mutex
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:32]: Given the creation of a new reactor requires a lot of the manager fields I created 2 methods there

File: node/pkg/processor/reactor/manager.go [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+				p.gsLock.Lock()
+				r = NewReactor[K](p.group, p.config, p.gs, func(reactor *ConsensusReactor[K], oldState, newState State) {
+					p.transitionHook(digest, reactor, oldState, newState)
+				})
+				p.gsLock.Unlock()
+				err := supervisor.Run(ctx, fmt.Sprintf("reactor-%s", digest.String()), r.Run)
+				if err != nil {
+					p.logger.Error("failed to spawn reactor routine", zap.Error(err), zap.Stringer("digest", digest))
+				}
+				p.reactors[digest] = r
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:171]: Moved into a method

File: node/pkg/processor/reactor/manager.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+				if p.storage != nil {
+					existingObservation, _, found, err := p.storage.GetSignedObservation(m.MessageId)
+					if err != nil {
+						p.logger.Warn("failed to check db for existing signed observation", zap.String("message_id", m.MessageId), zap.Stringer("digest", digest), zap.Error(err))
+					}
+					if found {
+						p.logger.Debug("ignoring confirmation - already in storage and no live reactor", zap.String("message_id", m.MessageId), zap.Stringer("digest_confirmed", digest), zap.Stringer("digest_stored", existingObservation.SigningMsg()))
+						continue
+					}
+				}
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:160]: turned this into a filter

File: node/pkg/processor/reactor/reactor.go [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+}
+
+func (c *ConsensusReactor[K]) Run(ctx context.Context) error {
+	c.logger = supervisor.Logger(ctx)
+	c.logger.With(zap.String("group", c.group))
+
+	c.stateMutex.Lock()
+	if c.state == StateFinalized {
+		c.stateMutex.Unlock()
+		return nil
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go:142]: Good catch

File: node/pkg/processor/reactor/manager.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+
+				p.gsLock.Lock()
+				r = NewReactor[K](p.group, p.config, p.gs, func(reactor *ConsensusReactor[K], oldState, newState State) {
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:163]: Only loading once to make sure this never becomes an issue.

File: node/pkg/processor/reactor/reactor.go [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+func (c *ConsensusReactor[K]) observationReceived(ctx context.Context, m *gossipv1.SignedObservation) {
+	c.stateMutex.Lock()
+	defer c.stateMutex.Unlock()
+
+	observationsReceivedTotal.WithLabelValues(c.group).Inc()
+
+	c.logger.Debug("received foreign observation", zap.Any("observation", m))
+
+	if c.state != StateObserved && c.state != StateUnobserved && c.state != StateQuorum && c.state != StateQuorumUnobserved && c.state != StateInitialized {
+		return
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go:297]: There are certain cases like double-observations that would be caught here. As this is not solely a transition function but also a handler I think we should not panic.

File: node/pkg/processor/reactor/reactor.go (hunk 1 above)
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go]: Moved to state

File: node/pkg/processor/reactor/reactor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jynnantonix) [reply node/pkg/processor/reactor/reactor.go:262]: I agree that putting them in a higher layer makes sense.  I'm fine with doing this in a follow-up change but if we end up cutting a release that doesn't have it we should let the community managers know ahead of time so that they can prepare for doing manual re-observations.

**Review by jynnantonix — DISMISSED** [review]
Nice! lgtm

File: node/pkg/processor/reactor/reactor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor.go:264]: Added them in the processor

**Comment by hendrikhofstadt** [comment]
> Any concern about the sheer number of reactors given thirty pyth messages per second?

Generally, no. Goroutines are very lightweight. However, I agree that we could reduce timeouts for pyth messages such that there is less load on p2p caused by unnecessary retransmissions. I.e. no retransmissions for unreliable messages at all.

**Review by jynnantonix — APPROVED** [review]

**Review by bruce-riley — APPROVED** [review]

**Comment by hendrikhofstadt** [comment]
@bruce-riley addressed the missing governor submission

**Review by bruce-riley — APPROVED** [review]

**Review by jynnantonix — APPROVED** [review]

**Comment by evan-gray** [comment]
@hendrikhofstadt saw you pushed yesterday, are you ready for re-review?

**Comment by evan-gray** [comment]
This has conflicts, so I'm backing to draft 🙂

File: node/pkg/processor/vaa_processor.go [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+		msgC:              msgC,
+		obsC:              obsC,
+		gossipSendC:       gossipSendC,
+		gst:               gst,
+		db:                db,
+		attestationEvents: attestationEvents,
+		governor:          g,
+		acct:              acct,
+		acctReadC:         acctReadC,
+	}
```
Comment (bruce-riley) [inline node/pkg/processor/vaa_processor.go:85]: This line is missing:
`pythnetVaas:       make(map[string]PythNetVaaEntry),`

**Review by jynnantonix — DISMISSED** [review]

**Review by bruce-riley — DISMISSED** [review]

**Review by panoel — DISMISSED** [review]
I only looked at the files in node/cmd/guardiand

**Comment by tbjump** [comment]
@hendrikhofstadt Do you maybe want to split out the "cleanup" changes (removing devNumGuardians, discordToken, ethRPC, wormchainLCD, VAA injection) into a separate PR? I think those are probably easy wins that could get merged sooner separate from the processor refactor. 

I started reviewing the code, and had a few initial questions: 
1. Is the goal of this change mostly to improve the code quality/readability, or is there new functionality, or is this a necessary change to add planned new functionality later? 
2. What about the sending of re-observation requests? Should that be part of the state diagram? 
3. Should accounting and governor be factored into this state machine? I think about accounting and governor as steps that happen before a message reaches the "Message Observation" transition. I could either see new states being added to this state machine or accountant and governor be completely de-coupled. 
4. What is the testing strategy for this change?

**Comment by hendrikhofstadt** [comment]
> @hendrikhofstadt Do you maybe want to split out the "cleanup" changes (removing devNumGuardians, discordToken, ethRPC, wormchainLCD, VAA injection) into a separate PR? I think those are probably easy wins that could get merged sooner separate from the processor refactor.
> 
> I started reviewing the code, and had a few initial questions:
> 
> 1. Is the goal of this change mostly to improve the code quality/readability, or is there new functionality, or is this a necessary change to add planned new functionality later?
> 2. What about the sending of re-observation requests? Should that be part of the state diagram?
> 3. Should accounting and governor be factored into this state machine? I think about accounting and governor as steps that happen before a message reaches the "Message Observation" transition. I could either see new states being added to this state machine or accountant and governor be completely de-coupled.
> 4. What is the testing strategy for this change?

@tbjump 
First of all I think it makes sense to factor out the changes you mentioned and I'll do that.

1. There are multiple reasons for this change: Overall Code quality and maintainability by breaking logic up in smaller more well defined components (particularly payload specific code from pure consensus), allowing consensus to be performed on different payloads without having to duplicate the reactor code by generalizing it (relevant for message accumulators and batches) and to improve performance by using multiple goroutines for consensus reactors
2. Reobservations just like governor and accounting are VAA payload specific elements of a workflow and not part of the raw consensus (i.e. agree on a digest). In the implementation this is reflected by these elements being handled by the VAA processor which itself sits above consensus and has it's own data flows.
3. See 2
4. While the refactoring done in this change will make testing much easier, this code is, like the reactor before, untested in CI. Tilt tests exercise the p2p network and reactors for functionality but not full correctness and handling edge-cases. The goal is to introduce strong test coverage in a follow-up PR.

**Comment by tbjump** [comment]
Thanks for the explanation. Looking at the previous BatchVAA PR (https://github.com/wormhole-foundation/wormhole/pull/1522) it makes more sense now.

**Comment by tbjump** [comment]
Please merge in https://github.com/wormhole-foundation/wormhole/pull/2351 and https://github.com/wormhole-foundation/wormhole/pull/2405

**Review by tbjump — CHANGES_REQUESTED** [review]
First off, I appreciate the effort to refactor the processor to make it easier to read and extend in the future. I left some inline comments and have some high-level feedback: 

* A significant refactor like this should come with extensive testing. The existing integration tests do not exercise enough codepaths. The existing code has been extensively tested in production.
* This refactor improves readability by more explicitly encoding the state machine and state machine transitions, but I also think that the usage of interfaces and generics decreases readability somewhat because the abstractions aren't perfect yet. Maybe readability could be further improved by stronger abstractions and better documentation of interfaces and responsibilities.
* A stated goal is to increase performance through concurrency. To me (I could well be wrong) it seems like that the per-message memory consumption is greatly increased (at least 2kb memory for the go routine) while the signature verification hotspot is still sequential. I don't know what the mainnet load expectations are for messages that are pending in consensus. We should do some benchmarks/profiling based on real-world loads to determine what's worth to optimize and if the optimizations are working. 
* There is currently no load-testing for Wormhole. I'd be reluctant to release such a significant change without load-testing or some other formalized testing plan.
* it's getting more and more difficult to track throughout the codebase who is writing to channels and who is reading to channels. I am not sure how to improve that, though. Any idea?

File: node/pkg/processor/vaa_processor.go [hunk 15]
```diff
@@ -0,0 +1,456 @@
...
+		zap.Any("vaa", v),
+		zap.String("bytes", hex.EncodeToString(m.Vaa)),
+		zap.String("message_id", v.MessageID()))
+
+	if err := p.storeSignedVAA(v); err != nil {
+		p.logger.Error("failed to store signed VAA", zap.Error(err))
+		return
+	}
+	p.attestationEvents.ReportVAAQuorum(v)
+}
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:316]: Vaa storage is very specific to VAAs and their type. Consensus is an independent concept which is why this lives here.

File: node/pkg/processor/reactor/reactor_signer.go [hunk 16]
```diff
@@ -0,0 +1,35 @@
+package reactor
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/reactor_signer.go:1]: Absolutely!

File: node/pkg/processor/vaa_processor.go [hunk 17]
```diff
@@ -0,0 +1,456 @@
...
+	"github.com/certusone/wormhole/node/pkg/processor/reactor"
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	"github.com/certusone/wormhole/node/pkg/reporter"
+	"github.com/certusone/wormhole/node/pkg/supervisor"
+	"github.com/wormhole-foundation/wormhole/sdk/vaa"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+type VAAReactor struct {
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:27]: Any idea for a better description. Reactor is terminology that I mainly picked up from the cosmos codebase. 
`VAAConsensusProcessor` is another option. The new docstring should help understand what it does.

File: node/pkg/processor/vaa_processor.go [hunk 18]
```diff
@@ -0,0 +1,456 @@
...
+			if p.acct != nil {
+				shouldPub, err := p.acct.SubmitObservation(k)
+				if err != nil {
+					return fmt.Errorf("acct: failed to process message `%s`: %w", k.MessageIDString(), err)
+				}
+				if !shouldPub {
+					continue
+				}
+			}
+
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:141]: Great catch!

File: node/pkg/processor/vaa_processor.go [hunk 19]
```diff
@@ -0,0 +1,456 @@
...
+				}
+
+				req := &gossipv1.ObservationRequest{
+					ChainId: uint32(observation.EmitterChain),
+					TxHash:  observation.TxHash.Bytes(),
+				}
+				if err := common.PostObservationRequest(p.obsvReqSendC, req); err != nil {
+					p.logger.Warn("failed to broadcast re-observation request", zap.Error(err))
+				}
+			})
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:216]: Refactored into a separate function. Reobservations are very speciic to VAA observations and are not part of "consensus on a digest" which is why I think they should be part of the vaa specific components.

File: node/pkg/processor/reactor/manager.go [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+				p.logger.Debug("failed to verify signed observation - dropping",
+					zap.Error(err),
+					zap.Stringer("digest", digest),
+					zap.String("signature", hex.EncodeToString(m.Signature)),
+					zap.String("addr", hex.EncodeToString(m.Addr)),
+				)
+				continue
+			}
+
+			r, err := p.loadOrCreateReactor(ctx, digest, gs, p.storageObservationFilter(m.MessageId))
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:138]: Tracking in another comment

File: node/pkg/processor/vaa_processor.go [hunk 21]
```diff
@@ -0,0 +1,456 @@
...
+				p.logger.Debug("ignoring observation since we already have a quorum VAA for it",
+					zap.Stringer("emitter_chain", k.EmitterChain),
+					zap.Stringer("emitter_address", k.EmitterAddress),
+					zap.Uint32("nonce", k.Nonce),
+					zap.Stringer("txhash", k.TxHash),
+					zap.Time("timestamp", k.Timestamp),
+					zap.String("message_id", existing.MessageID()),
+				)
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:154]: Yes, let's open an issue. The goal here is to develop something that has mostly identical functionality as before.

File: node/pkg/processor/reactor/manager.go [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+	// setC is a channel of guardian set updates
+	setC <-chan *common.GuardianSet
+
+	// reactors are the current live reactors. This list may contain prepared reactors in StateInitialized. This field
+	// may only be touched by removeReactor and loadOrCreateReactor.
+	reactors     map[ethcommon.Hash]*ConsensusReactor[K]
+	reactorsLock sync.Mutex
+
+	// gs is the currently valid guardian set.
+	gs atomic.Pointer[common.GuardianSet]
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:42]: Good point!

**Comment by hendrikhofstadt** [comment]
@tbjump Thanks for the exhaustive feedback!

First of all this PR itself lacks a lot of the context around the testing strategy for this change. As mentioned it's absolutely necessary to add test coverage which I've started to work on.
However the goal was never to deploy this straight into production. A change of this magnitude requires potentially multiple months of testing in a prod environment. Some folks had started to work on a testing strategy to roll this into testnet and potentially recommending a rollout on a minority to superminority of guardians before handing changes to a supermajority. Given these changes are backwards compatible this should be possible although timeouts will have to be adapted.

Overall I've spent a good amount thinking about how to break the reactor up from a highly sequential purpose-specific monolith into small clearly purpose built components that can be layered. 
Given Go's intentionally limited Generics design, there are limits to how well certain pieces can be abstracted.

The main goal is to break consensus from VAAs specifically but more into "WH consensus on a payload".
**Reactors** do this on a single payload of type _K_ (which is _Observable_ a intentionally generic type), **Managers** handle the lifecycle of multiple reactors of type _K_ and dispatch packets between them, **Processors** (VAAReactor is poor naming in this case) implement payload-specific logic such as VAA reobservations.

I would be curious on your ideas for further abstractions to improve readability and usability as I personally was not able to find elements to break up into even smaller components or abstract away without significantly increasing complexity.

File: node/pkg/processor/reactor/manager.go [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+	// group of the manager. This is a label for this manager and all associated reactors.
+	group string
+	// observationC is a channel of observed emitted messages
+	observationC <-chan K
+
+	// confirmationC is a channel of inbound decoded observations from p2p
+	confirmationC <-chan *gossipv1.SignedObservation
+
+	// setC is a channel of guardian set updates
+	setC <-chan *common.GuardianSet
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:28]: Strongly agree!

File: node/pkg/processor/reactor/manager.go [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+func (p *Manager[K]) transitionHook(digest ethcommon.Hash, reactor *ConsensusReactor[K], oldState, newState State) {
+	switch newState {
+	case StateQuorum:
+		p.logger.Debug("reactor reached quorum", zap.Stringer("digest", digest), zap.String("message_id", reactor.Observation().MessageID()))
+
+		// Store in the database
+		if p.storage != nil {
+			err := p.storage.StoreSignedObservation(reactor.Observation(), reactor.VAASignatures())
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go]: I think the storage adapter should handle that. This does not pass in a context which I interpret as this should not do IO. So I will clearly document that this should never block. Afaik the DB we are using does not flush on write so this should be fine but i'll double-check and document.

File: node/pkg/processor/reactor/reactor.go [hunk 25]
```diff
@@ -0,0 +1,535 @@
...
+		group: group,
+		state: reactorState[K]{
+			currentState: StateInitialized,
+			signatures:   map[ethcommon.Address][]byte{},
+		},
+		gs:                        gs,
+		stateTransitionChan:       s,
+		config:                    config,
+		foreignObservationChannel: make(chan *gossipv1.SignedObservation, 10),
+		observationChannel:        make(chan K, 10),
```
Comment (tbjump) [inline node/pkg/processor/reactor/reactor.go:144]: the `foreignObservationChannel` should probably be bigger (expected up to 18 concurrently) and the `observationChannel` smaller (only expected one at a time).

File: node/pkg/processor/vaa_processor.go [hunk 26]
```diff
@@ -0,0 +1,473 @@
...
+		UnobservedTimeout:   24 * time.Hour,
+		Signer:              reactor.NewEcdsaKeySigner(gk),
+		NetworkAdapter:      reactor.NewChannelNetworkAdapter(gossipSendC),
+	}, r, r)
+	r.manager = manager
+
+	return r
+}
+
+func (p *VAAReactor) Run(ctx context.Context) error {
```
Comment (tbjump) [inline node/pkg/processor/vaa_processor.go:101]: What do you think about doing the processing of each event (big select statement below) in it's own goroutine?

File: node/pkg/processor/vaa_processor.go [hunk 27]
```diff
@@ -0,0 +1,473 @@
...
+	defer govTimer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if p.acct != nil {
+				p.acct.Close()
+			}
+			return ctx.Err()
+		case k := <-p.msgC:
```
Comment (tbjump) [inline node/pkg/processor/vaa_processor.go:126]: idea: put token bridge messages (or anything that's covered by governor or accountant) into its own channel and process concurrently; drop them if said channel is full to not hold up processing of other messages.

File: node/pkg/processor/reactor/manager.go [hunk 28]
```diff
@@ -0,0 +1,236 @@
...
+
+// Manager handles the creation and maintenance of reactors for incoming local and foreign observations
+type Manager[K Observation] struct {
+	// group of the manager. This is a label for this manager and all associated reactors.
+	group string
+	// observationC is a channel of observed emitted messages
+	observationC <-chan K
+
+	// confirmationC is a channel of inbound decoded observations from p2p
+	confirmationC <-chan *gossipv1.SignedObservation
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go:25]: Given protos can be extended I think we will stick to gossipv1 forever

File: node/pkg/processor/reactor/manager.go [hunk 29]
```diff
@@ -0,0 +1,236 @@
...
+	if reactor, exists := p.reactors[digest]; exists {
+		p.logger.Debug("found existing reactor", zap.Stringer("digest", digest))
+		r = reactor
+	} else {
+		if filter != nil && !filter(digest) {
+			return nil, fmt.Errorf("filter prevented creation of new reactor")
+		}
+
+		p.logger.Debug("creating new reactor", zap.Stringer("digest", digest))
+		r = NewReactor[K](p.group, p.config, gs, func(reactor *ConsensusReactor[K], oldState, newState State) {
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/reactor/manager.go]: I'll move QoS changes into a separate PR (or commit) and address: reactor lifetime, per-guardian limits, etc

**Review by tbjump — CHANGES_REQUESTED** [review]
Left a few more comments. My main outstanding issues are: 

* We don't yet know the Memory and CPU implications of this change. I'm working on a PR that will allow benchmarking of the Guardian. We should rebase this change onto my benchmark PR and compare it vs. main.
* Increased memory consumption per message, combined with messages being kept around much longer after quorum (QuorumTimeout = 24h) could lead to massively increased memory consumption. Looking at Wormhole daily message volume (2M/day), memory required will be at least around 4 GB. If Wormhole message volume increases 100x, that could exhaust the memory of Guardians. We'll learn more exact numbers with the benchmarking, but regardless I'd suggest that 
  * reactors should reach `StateFinalized` after getting 19 signatures, even if that's before the timeout
  * If the number of reactors reaches a threshold, space should be made available by advancing reactors from `StateQuorum` to `StateFinalized` even if that happens before the timeout. 
* Unit tests look good, but still want to look at line-by-line test coverage (@SEJeff is working on it).

File: node/pkg/processor/vaa_processor.go [hunk 30]
```diff
@@ -0,0 +1,522 @@
...
+		acct:              acct,
+		acctReadC:         acctReadC,
+		pythnetVaas:       map[string]PythNetVaaEntry{},
+		gst:               gst,
+	}
+
+	manager := reactor.NewManager[*vaaObservation]("vaa", obsC, obsvC, gst, reactor.Config{
+		RetransmitFrequency: 5 * time.Minute,
+		QuorumGracePeriod:   2 * time.Minute,
+		QuorumTimeout:       24 * time.Hour,
```
[the remark this answers is not in the document]
Comment (hendrikhofstadt) [reply node/pkg/processor/vaa_processor.go:93]: Agree, this could easily just be 2 minutes.

File: node/pkg/processor/reactor/reactor.go [hunk 31]
```diff
@@ -0,0 +1,544 @@
...
+	c.state.mutex.Lock()
+	defer c.state.mutex.Unlock()
+	var sigs []*vaa.Signature
+	for i, a := range c.gs.Keys {
+		s, ok := c.state.signatures[a]
+
+		if ok {
+			var bs [65]byte
+			if n := copy(bs[:], s); n != 65 {
+				panic(fmt.Sprintf("invalid sig len: %d", n))
```
Comment (tbjump) [inline node/pkg/processor/reactor/reactor.go:244]: please comment under which conditions this panic could be reached and why that should be impossible. consider adding unit tests.

File: node/pkg/processor/reactor/reactor.go [hunk 32]
```diff
@@ -0,0 +1,544 @@
...
+	c.logger.Debug("state transition", zap.String("from", string(c.state.currentState)), zap.String("to", string(to)))
+	previousState := c.state.currentState
+	c.state.currentState = to
+
+	if c.stateTransitionChan != nil {
+		transition := &StateTransition[K]{FromState: previousState, ToState: to, Reactor: c}
+		select {
+		case c.stateTransitionChan <- transition:
+		default:
+			c.logger.Warn("state transition channel full; dropped update", zap.Any("transition", transition))
```
Comment (tbjump) [inline node/pkg/processor/reactor/reactor.go:472]: Not covered by unit tests. 
Please explain why we'd want to drop transitions here.

File: node/pkg/processor/reactor/reactor.go [hunk 33]
```diff
@@ -0,0 +1,544 @@
...
+		observationsFailedTotal.WithLabelValues(group, "pubkey_mismatch").Inc()
+		return fmt.Errorf("address does not match pubkey: %s != %s", theirAddr.Hex(), sigPub.Hex())
+	}
+
+	// Verify that m.Addr is included in the guardian set. If it's not, drop the message. In case it's us
+	// who have the outdated guardian set, we'll just wait for the message to be retransmitted eventually.
+	_, ok := gs.KeyIndex(theirAddr)
+	if !ok {
+		observationsFailedTotal.WithLabelValues(group, "unknown_guardian").Inc()
+		return fmt.Errorf("unknown guardian: %s", theirAddr.Hex())
```
Comment (tbjump) [inline node/pkg/processor/reactor/reactor.go:540]: not covered by unit tests, but should be.

**Review by tbjump — CHANGES_REQUESTED** [review]
Added a few more comments after reviewing test coverage.

**Comment by tbjump** [comment]
Closing in favor of https://github.com/wormhole-foundation/wormhole/pull/3140
